### PR TITLE
Nathan Lambertson

### PR DESCRIPTION
**Pull request by lambo1986**
Arrrggggg

### CODE REVIEW

File: lib/exhibit.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class Exhibit
+  attr_reader :name,
+              :cost
+  def initialize(name_and_cost_hash)
```
Comment (KatBrandt) [inline lib/exhibit.rb:4]: Good job taking in the hash and reading the attributes from it. In terms of the variable name used, I would just use something more generic like `exhibit_data` or just `data`.

File: lib/patron.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+              :interests
+  def initialize(name, monies_count)
+    @name = name
+    @spending_money = monies_count
+    @interests = []
+  end
+
+  def add_interest(interest)
+    @interests << interest
+  end
```
Comment (KatBrandt) [inline lib/patron.rb:13]: Great job with this class, looks good.

File: lib/museum.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  def admit(patron)
+    @patrons << patron
+  end
+
+  def recommend_exhibits(patron)
+    recommended_exhibits = []
+    patron.interests. each do |interest|
+      @exhibits.each do |exhibit|
+        recommended_exhibits << exhibit if exhibit.name.include?(interest)
```
Comment (KatBrandt) [inline lib/museum.rb:25]: Great job here!!! Using nested iteration and then figuring out what to compare. Nice work.

File: lib/museum.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  def patrons_by_exhibit_interest
+  interest_hash = Hash.new
+    @exhibits.each do |exhibit| #trouble getting all patrons into respective keys
+      @patrons.each do |patron| #wasted a lot of time :( can't get more than one patron to go to hash value
+      interest_hash[exhibit] = [patron] if patron.interests.include?(exhibit.name)
+      end
+    end
+    interest_hash
+  end
```
Comment (KatBrandt) [inline lib/museum.rb:39]: Yeah, this is a tough one! The problem you're encountering is that on line 35, you are resetting the key's value everytime you call it. So if that exhibit was already a key in the hash with a patron in an array, this line would then override that and replace the value with the new patron. 
So we need a way to _either_ create the array with the patron if that key doesn't already exist, OR, shovel that patron into the exisiting array if it already exists. Hashes can have default values as well which might help with this one. Definitely keep working on it!

File: lib/museum.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  def ticket_lottery_contestants(exhibit)
+    contestants = []
+    @patrons.each do |patron|
+      contestants << patron if patron.spending_money < exhibit.cost
```
Comment (KatBrandt) [inline lib/museum.rb:44]: Close! But we also want to make sure the patron also has an interest in that exhibit too. So just adding another conditional will help with this.

File: lib/museum.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    end
+    @lottery_contestants << contestants
```
Comment (KatBrandt) [inline lib/museum.rb:46]: I'm curious why you choose to make an instance variable here. Usually, if we can just calculate the data dynamically within the method, then we don't need to store the results in an instance variable.

File: lib/museum.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    contestants
+  end
+
+  def draw_lottery_winner
+    winner = @lottery_contestants.sample
```
Comment (KatBrandt) [inline lib/museum.rb:51]: I see why you stored it in an instance variable, to use it here. However, that's not necessary. You could have just called `winner = ticket_lottery_contestants(exhibit).sample`. Then this would call that method for you, return the results, and then pick the winner. :)

File: lib/museum.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    winner.name
+  end
+end
```
Comment (KatBrandt) [inline lib/museum.rb:54]: And then you didn't get to the `announce_lottery_winner` method it seems?

File: spec/museum_spec.rb [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    patron_1 = Patron.new("Bob", 0)
+    dmns.admit(patron_1)
+
+    expect(dmns.patrons.count).to eq(1)
+  end
+
+  it "has an attribute for exhibits" do
+    dmns = Museum.new("Denver Museum of Nature and Science")
+
+    expect(dmns.exhibits).to be_a(Array)
```
Comment (KatBrandt) [inline spec/museum_spec.rb:41]: For this one, we also want to test that when we first initialize a Museum object, that this attribute is not only an Array, but an empty array. So could just have line 41 say `expect(dmns.exhibits).to eq([])`.

File: spec/museum_spec.rb [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  end
+
+  it "has a method to add exhibit instances to exhibits attribute" do
+    dmns = Museum.new("Denver Museum of Nature and Science")
+    gems_and_minerals = Exhibit.new({name: "Gems and Minerals", cost: 0})
+    dead_sea_scrolls = Exhibit.new({name: "Dead Sea Scrolls", cost: 10})
+    imax = Exhibit.new({name: "IMAX",cost: 15})
+    dmns.add_exhibit(gems_and_minerals)
```
Comment (KatBrandt) [inline spec/museum_spec.rb:49]: Feel free to add spacing in your tests for better readability. I like to add a white line between my setup, the execution, and the assertion.

File: spec/museum_spec.rb [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    dmns.add_exhibit(dead_sea_scrolls)
+    dmns.add_exhibit(imax)
+    expect(dmns.exhibits.count).to eq(3)
```
Comment (KatBrandt) [inline spec/museum_spec.rb:52]: We want to test more explicitly. I'd want to make sure that not only are there 3 elements in my array now, but that they are specifically the objects I passed into the `add_exhibit` method, so line 52 would be more `expect(dmns.exhibits).to eq ([gems_and_minerals, dead_sea_scrolls, imax])`

File: spec/museum_spec.rb [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    dmns.add_exhibit(gems_and_minerals)
+    dmns.add_exhibit(dead_sea_scrolls)
+    dmns.add_exhibit(imax)
+    patron_1 = Patron.new("Bob", 20)
+    patron_1.add_interest("Dead Sea Scrolls")
+    patron_1.add_interest("Gems and Minerals")
+    patron_2 = Patron.new("Sally", 20)
+    patron_2.add_interest("IMAX")
+    expect(dmns.recommend_exhibits(patron_1).count).to eq(2)
+    expect(dmns.recommend_exhibits(patron_2).count).to eq(1)
```
Comment (KatBrandt) [inline spec/museum_spec.rb:69]: Same thing here, how we can test more explicitly?

**Review by KatBrandt — COMMENTED** [review]
Overall you're doing great Nate! I'd recommend meeting up with a mentor/rock/student support, to review my comments and get more guidance on writing explicit tests. Your work with enumerables is looking good. You were so close on the one patrons by exhibit method!
